x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.ml.inference.utils;
+
+import org.elasticsearch.common.CheckedFunction;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Collection of helper methods. Similar to CompressedXContent, but this utilizes GZIP for parity with the native compression
+ */
+public final class ToXContentCompressor {
+    private static final int BUFFER_SIZE = 4096;
+
+    private ToXContentCompressor() {}
+
+    public static <T extends ToXContentObject> String deflate(T objectToCompress) throws IOException {
+        BytesReference reference = XContentHelper.toXContent(objectToCompress, XContentType.JSON, false);
```

Comment:
why JSON? A binary encoding might be better suited, we support CBOR and SMILE ootb. It might not make a difference in terms of size but in terms of parsing it back (runtime).

Reply:
The reason for JSON was future interop with native. I think it would be nice to have a definition that is trivially supported by the native process so we can simply pass the compressed strings without any processing.

Reply:
:+1:  I forgot that this needs native support, at least SMILE has no good implementation for C++. LBNL parsing should not be the bottleneck.